### ISSUE TITLE
Mongoose Disconnect/reconnect event Issue

### ISSUE THREAD

**Issue by manoj97738** [Automattic/mongoose#3183]
I am currently using mongoose@4.0.4, whenever I manually disconnect MongoDb, mongoose  disconnect/reconnect event are not fired corrected, reconnect event is fired and disconnect is not being fired every time the MongoDb  is disconnect/started manually. My problem is to handle MongoDb connect and disconnected status so that for every request made to server I would see the status of the MongoDb connection/disconnection and send appropriate response(may be 500 code) because the request hangs up in middle if MongoDb is not connected.
To see if this was the version issue I stated version 3.8.x and went to 4.0.4 to see where the events are fired properly. I was able to catch reconnect event from version 4.0.x but not the disconnect event.
Basically wanted to know which version i can use where connect/disconnect events are fired properly or way to handle the issue.

**Comment by vkarpov15** [comment]
Try upgrading to 4.0.8, there where a few related issues in some earlier 4.0.x versions but I don't remember the exact number. If they still don't work properly in 4.0.8, that's a bad bug :)

**Comment by manoj97738** [comment]
Thanks for the Quick reply Sir, tried v 4.0.8 still the same issue.

**Comment by vkarpov15** [comment]
And can you show me some code that reproduces this?

**Comment by manoj97738** [comment]
Thank SIr,

please find the below code I have used (node v10.0.32)

var db = mongoose.connection;
var connected;
db.on('open', function (ref) {
    connected = true;
    console.log('open connection to mongo server.');
});

db.on('connected', function (ref) {
    global.mongo_conn=true;
	console.log('Connected connection to mongo server.');
});

db.on('disconnected', function (ref) {
	connected = false;
    console.log('disconnected connection.');
});
db.on('disconnect', function (err) {
    console.log('Error...disconnect', err);
});
db.on('connecting', function (ref) {
	connected = false;
    console.log('connecting.');
});

db.on('close', function (ref) {
	global.mongo_conn=false;
    console.log('close connection.');
    connect();
});

db.on('error', function (ref) {
	connected = false;
    console.log('Error connection.');
    //mongoose.disconnect();
	global.mongo_conn=false;
});

db.on('reconnected', function () {
    global.mongo_conn=true;
	console.log('MongoDB reconnected!');
});
db.on('reconnecting', function () {
    global.mongo_conn=true;
	console.log('reconnecting!');
});
function connect() {
    mongoose.connect(config.mongo.uri, opts);
}

connect();
